### DISCUSSION TITLE
Massive link headers causing site failures

### DISCUSSION THREAD

**Discussion by kevinrenskers** [sveltejs/kit#6519] [Q&A] [answered]
Today I ran into a really strange issue: when I opened https://www.critical-notes.com/campaigns/130/characters/4955 I got an error 502 Bad Gateway page from Nginx, but I was able to load https://www.critical-notes.com/campaigns/130/characters just fine, and navigate to a character. But I couldn't reload, or I would get the 502 again. 

After some searching I found out that the link header is so big that nginx just refuses to deal with it:

`2022/09/01 22:36:39 [error] 3352048#3352048: *4228330 upstream sent too big header while reading response header from upstream, client: 35.237.4.214, server: www.critical-notes.com, request: "GET /campaigns/130/characters/4955 HTTP/1.1", upstream: "http://127.0.0.1:3000/campaigns/130/characters/4955", host: "www.critical-notes.com"`

And looking at the header, there are a LOT of JS files in there, it's really rather big. 

I did managed to solve it by adding this to my nginx site config:

```
proxy_buffer_size 16k;
proxy_buffers 4 16k;
proxy_busy_buffers_size 16k;
```

But is this to be expected? What's that link header doing, does it need to be so big? Is this documented somewhere? And in general, is there any documentation on hosting this yourself? I am currently using adapter-node and nginx and it works (until it doesn't). Would love to get some official docs on this.

**Comment by tmsnarr** [comment]
I ran into the same issue and @kevinrenskers solution of adding those options to the top line of my 443 server block worked for me too. I'm on version 1.0.0-next.465 and use the node-adapter. 

I have a route that uses a slug that couldn't be reloaded in production, but could be navigated to from parent directory via client-side nav. When reloading the slug page, the 502 would show up and you'd have to push back on the browser and then refresh again to get the site working. I found the exact same error as above in my NGINX logs.

**Comment by benmccann** [comment]
Can you see if setting this option helps: https://github.com/sveltejs/kit/issues/2612#issuecomment-1208715755

If it doesn't, would it be possible to share your code? I'm curious why your page would be using so many JS files

**Reply by kevinrenskers** [reply]
I can't share my code. But if you fetch https://www.critical-notes.com/campaigns/130/characters/4955 using something like Paw or Postman or just curl, you can see the link header:

`<../../../_app/immutable/assets/_layout-20f85abe.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/assets/FormManyPicker-b25275b8.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/assets/_layout-083ac408.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/assets/_layout-73abed0f.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/assets/CharacterRow-3ec706d9.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/assets/_page-02742be2.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/assets/FactionRow-c76dd16b.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/assets/LocationRow-3301da6b.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/assets/QuestRow-8356e42e.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/assets/LootRow-409e2f7b.css>; rel="preload";as="style"; nopush, <../../../_app/immutable/start-507ae4f3.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/index-b5628a3e.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/singletons-a828871f.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/control-a6874251.js>; rel="modulepreload"; nopush, <../../../_app/immutable/components/pages/_layout.svelte-0208a909.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/stores-1be62249.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/store-f39a0cb8.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/websockets-b585008c.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/navigation-e65b3ca9.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/config-ad25c1b2.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/FormManyPicker.svelte_svelte_type_style_lang-746b999e.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/index-b1d5cfe9.js>; rel="modulepreload"; nopush, <../../../_app/immutable/modules/pages/_layout.ts-901ed97a.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/api-773ae5c1.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/index-215d9995.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/_layout-635aad8e.js>; rel="modulepreload"; nopush, <../../../_app/immutable/components/pages/campaigns/_layout.svelte-b36df9ef.js>; rel="modulepreload"; nopush, <../../../_app/immutable/modules/pages/campaigns/_layout.ts-e1405392.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/utils-c8ee61ea.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/_layout-83150984.js>; rel="modulepreload"; nopush, <../../../_app/immutable/components/pages/campaigns/_campaignId_/_layout.svelte-0cdca5b3.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/Navbar-905577b6.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/SegmentedButton-6e24b5e5.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/ChoiceModal-7fd1f760.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/NavLink-f7cf7a76.js>; rel="modulepreload"; nopush, <../../../_app/immutable/modules/pages/campaigns/_campaignId_/_layout.ts-20986714.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/_layout-369b7dc7.js>; rel="modulepreload"; nopush, <../../../_app/immutable/components/pages/campaigns/_campaignId_/characters/_layout.svelte-263444b9.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/FormManyPickers-064f33cf.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/fetch-e80f3fa5.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/CharacterRow-643d442b.js>; rel="modulepreload"; nopush, <../../../_app/immutable/modules/pages/campaigns/_campaignId_/characters/_layout.ts-18fc60e2.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/_layout-6a05cae1.js>; rel="modulepreload"; nopush, <../../../_app/immutable/components/pages/campaigns/_campaignId_/characters/_characterId_/_page.svelte-66afafbc.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/Markdown-53073373.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/DeleteModal-a0dbeb35.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/DetailsHeader-2994dbf6.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/Notes-79d8c48d.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/FactionRow-f5985ab4.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/LocationRow-0b7ee234.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/QuestRow-73b90e5b.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/LootRow-0a08c23a.js>; rel="modulepreload"; nopush, <../../../_app/immutable/modules/pages/campaigns/_campaignId_/characters/_characterId_/_page.ts-1df226df.js>; rel="modulepreload"; nopush, <../../../_app/immutable/chunks/_page-6eb6b2c3.js>; rel="modulepreload"; nopush`

I'm a bit afraid of using an experimental option in my production site, and don't have time to set up a staging server.

**Reply by benmccann** [reply]
I'm far less optimistic the option will really help now having seen the header. That being said, you could easily test the option locally. You might not know if it solved this issue without putting nginx in front of it, but you could pretty easily see that it's not going to break anything. We're about to turn it on for everyone and make it not experimental. The only real issue we need to fix first is that if you change Svelte compiler options during dev mode it won't pick that up, but it should be pretty safe to use in production.

We could offer some type of option to disable or reduce the link header, but I think we're much better off trying to reduce the number of JS files being loaded. The link header tells the browser that this is all the JS you will need for the page so it can get to work loading it as soon as possible. Disabling the option will make your site slower especially since there are so many JS files to load.

There are a couple things I noticed. One is that it sort of looks like you might have five layouts being used on the page? It's hard to tell just from the compiled code without seeing the source, but there's really a lot of layout-related files and I'd like to understand what's going on there. The other thing is that every menu item seems to have its own Svelte component. It'd probably help to refactor that such that there's just a generic MenuRow.svelte. I also wonder if there's something Vite could do to bundle these together into a single file as I'm assuming they're always used together

**Reply by kevinrenskers** [reply]
> One is that it sort of looks like you might have five layouts being used on the page? It's hard to tell just from the compiled code without seeing the source, but there's really a lot of layout-related files and I'd like to understand what's going on there.

There are 4 layouts involved here.
Characters -> the single campaign -> the list of campaigns -> root layout.

The character detail page is slotted into the characters list layout so you have the list of characters on the left, and then the details `<slot />` on the right. The characters page is slotted into a campaign layout, which adds the campaign submenu on the far left. All that is wrapped in the campaigns list layout which is literally an empty `<slot />` with nothing else, but it's needed because I have a `+layout.ts` file on that level which actually loads all the campaigns from the server. And finally the root layout has the overal top menu, user avatar, user loading from the server, etc.

![Screen Shot 2022-09-11 at 20 44 24](https://user-images.githubusercontent.com/229384/189543977-c0ac6b5b-1b4a-47e1-a178-0eafbc4cf528.png)

> The other thing is that every menu item seems to have its own Svelte component. It'd probably help to refactor that such that there's just a generic MenuRow.svelte.

No, the characters page shows linked quests, locations and factions, which all have their own component.

![Screen Shot 2022-09-11 at 20 45 55](https://user-images.githubusercontent.com/229384/189544029-4e172f71-535d-4bd9-98f2-3d40f3cd3185.png)

Considering Svelte works best when stuff is broken up into components.. yeah there are a bunch of them on the page.

> I also wonder if there's something Vite could do to bundle these together into a single file as I'm assuming they're always used together

Agreed. I'm actually kinda surprised all these little JS and CSS are not bundled.

**Reply by tmsnarr** [reply]
I'm also using nested layouts, but just 2. With this structure, `routes -> player -> [id]`, I have a layout in `routes` and one in `[id]` (I get the `502` from NGINX when reloading `[id]`). I'm importing 13 small- to medium-sized components in `[pid]/+page.svelte`, as well as a good amount of other JS functions from my utils.js and libraries like d3 and lodash. I'm also importing a few small components in `[pid]/+layout.svelte`.

**Comment by jesperordrup** [comment]
This should be reported as an issue ?

**Reply by kevinrenskers** [reply]
https://github.com/sveltejs/kit/issues/6790

**Comment by Rich-Harris** [comment]
Per https://github.com/sveltejs/kit/issues/6790#issuecomment-1254205734, this is easy to fix in a `handle` hook:

```js
export function handle({ event, resolve }) {
  const response = await resolve(event);
  response.headers.delete('link');
  return response;
}
```

**Reply by kristjanmar** [reply]
What about if we want to get rid of the link headers, but still revert to the previous behavior of having `<link rel="modulepreload">` in the html? Is that possible?

**Reply by obgibson** [reply]
where does this handle function go? Which file?

**Reply by aloker** [reply]
It should go into hooks.server.js (or .ts)

See https://kit.svelte.dev/docs/hooks

**Comment by Kamyil** [comment]
Yeah, we had to increase the proxy buffer in our NGINX config in order to fix this issue, since link header becomes quite too long. 

```
 # Needed because sveltekit generates long "link" header for preloading
        # https://www.getpagespeed.com/server-setup/nginx/tuning-proxy_buffer_size-in-nginx
        proxy_buffering off;
        proxy_buffer_size 32k;
        proxy_busy_buffers_size 40k;
        proxy_buffers 128 4k;     
```

We also saw that preload links are also represented in the form of HTML tags inside the <head> element. 
Do you know maybe why do we need link as headers if we already have them as HTML elements?

**Reply by lecramr** [reply]
Ever figured your last question out? Are they necessary? :)

**Reply by Kamyil** [reply]
Not sure. Probably SvelteKit relies on it internally in some way I presume

**Reply by diramazioni** [reply]
I presume it's cache related?

**Comment by teemingc** [comment, marked as answer]
I've opened https://github.com/sveltejs/kit/pull/15939 to disable `Link` header preloading as the default
